src/Control/Scanl.hs

```diff
@@ -0,0 +1,129 @@
+{-| This module provides efficient and streaming left map-with-accumulator that you can combine
+    using 'Applicative' style.
+
+    Import this module qualified to avoid clashing with the Prelude:
+
+>>> import qualified Control.Scanl as SL
+-}
+
+{-# LANGUAGE BangPatterns              #-}
+{-# LANGUAGE CPP                       #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE FlexibleContexts          #-}
+{-# LANGUAGE RankNTypes                #-}
+{-# LANGUAGE TupleSections             #-}
+
+module Control.Scanl (
+    -- * Scan Types
+      Scan(..)
+    , ScanM(..)
+
+    , prescan
+    , postscan
+
+    -- * Utilities
+    -- $utilities
+    , premap
+    ) where
+
+import Control.Arrow
+import Control.Category
+import Control.Foldl hiding (premap, prescan, postscan)
+import Control.Foldl.Internal (Pair(..))
+import Data.Profunctor
+import Prelude hiding ((.), id)
+
+{-| Efficient representation of a left map-with-accumulator that preserves the
+    scan's step function and initial accumulator.
+
+    This allows the 'Applicative' instance to assemble derived folds that
+    traverse the container only once
+
+    A \''Scan' a b\' processes elements of type __a__ replacing each with a
+    value of type __b__.
+-}
+data Scan a b
+  -- | @Scan @ @ step @ @ initial @
+  = forall x. Scan (x -> a -> (x, b)) x
+
+instance Functor (Scan a) where
+    fmap f (Scan step begin) = Scan (\x a -> f <$> step x a) begin
+    {-# INLINE fmap #-}
+
+instance Applicative (Scan a) where
+    pure b    = Scan (\() _ -> ((), b)) ()
+    {-# INLINE pure #-}
+
+    (Scan stepL beginL) <*> (Scan stepR beginR) =
+        let step (Pair xL xR) a = ((Pair xL' xR'), (bL bR))
+              where (xL', bL) = stepL xL a
+                    (xR', bR) = stepR xR a
+            begin = Pair beginL beginR
+        in  Scan step begin
+    {-# INLINE (<*>) #-}
+
+instance Profunctor Scan where
+    lmap = premap
+    rmap = fmap
+
+instance Category Scan where
+    id = Scan (\()-> \a -> ((), a)) ()
+    {-# INLINE id #-}
+    (Scan s2 b2) . (Scan s1 b1) = Scan step (Pair b1 b2)
+        where step (Pair xL xR) a = (Pair xL' xR', c)
+                where (xL', b) = s1 xL a
+                      (xR', c) = s2 xR b
+    {-# INLINE (.) #-}
+
+instance Arrow Scan where
+    arr f = Scan (\()-> \a -> ((), f a)) ()
+    {-# INLINE arr #-}
+    first  (Scan step begin) = Scan (\x (a,b) -> fmap (,b) $ step x a) begin
+    {-# INLINE first #-}
+    second (Scan step begin) = Scan (\x (b,a) -> fmap (b,) $ step x a) begin
+    {-# INLINE second #-}
+
+
+#if MIN_VERSION_base(4,8,0)
```

Comment:
What is the reason for this condition on the version of `base`?

Reply:
I cargo culted from Fold